mlflow/types/schema.py

```diff
@@ -408,3 +408,140 @@ def __eq__(self, other) -> bool:
 
     def __repr__(self) -> str:
         return repr(self.inputs)
+
+
+class ParamSpec:
+    """
+    Specification used to represent parameters for the model.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        type: str,  # pylint: disable=redefined-builtin
+        default: Any,
+    ):
+        self._name = str(name)
+        self._type = str(type)
+        self._default = default
+        self._check_default_type()
```

Comment:
It seems like we don't do any validation on these types (let me know if I'm missing something).

For example, if I passed in an instance of `MlflowClient` to `ParamSpec`, there wouldn't be any exception thrown, even though it would be impossible to specify an instance of `MlflowClient` as a parameter at inference time when serving the model using `mlflow models serve` / Databricks Model Serving.

I think the general problem is that `ParamSpec` seems to be allowing parameters to be defined with types that we can't support using the MLflow model server. Even the `tuple` type, which we test for, doesn't have a JSON representation and couldn't be used with the MLflow model server.

Finally, I think it might be a bit strange that we're using Python type names for the `type` attribute of `ParamSpec` and validating on them here: https://github.com/mlflow/mlflow/pull/8563/files#diff-c5c0f3e73e332718372ebf06602d3f6caa0a631cc8e7987a0919250f10cfa396R430, since these don't match the MLflow data type names defined in https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.DataType.

Here's what I think we should do:

- For `type`, only accept values defined in https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.DataType or (indicates a scalar parameter) a single-element Python list containing an instance of https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.DataType (indicates a list parameter)

I don't see an immediate need for more complex parameter structures beyond scalar and list. Let me know if y'all have questions or conflicting opinions here :)

cc @harupy @BenWilson2

Reply:
@serena-ruan I updated the proposal a bit ^. Lmk what you think.

Reply:
https://huggingface.co/docs/transformers/main_classes/text_generation#transformers.GenerationConfig.exponential_decay_length_penalty This is an example of tuple type parameter, basically now the `type` doesn't make any constraint, and it doesn't constraint on the inner types as well. So the question is what types should we support in general? https://huggingface.co/docs/transformers/main_classes/text_generation#transformers.GenerationConfig.constraints this is a list of `Constraint` class, and if we only limit to DataType, then there are parameters acceptable by the model but not acceptable by us, which makes this functionality only half-implemented. 
We might be able to detect allowable types for the current version of transformers, but to make `params` useful to all PyFunc models in general, it's hard to define allowable types.

Reply:
We need to be consistent with `TensorSpec` and `ColSpec`, and we also need to make sure that all of the types supported by `ParamSpec` are supported by the MLflow Model Server.

If the transformers flavor needs to support tuples, then IMO we should define conversion logic from List -> Tuple inside the transformers pyfunc wrapper class.

Reply:
Why do we need to be consistent with `TensorSpec` and `ColSpec`? I thought both of them and `DataType` are supposed to define data schemas, and the type should be convertible between pandas, numpy and spark. But `params` could be random value passed to the model directly e.g.https://huggingface.co/docs/transformers/main_classes/text_generation#transformers.GenerationConfig.constraints. Though I'm not sure how often people uses this parameter, but limiting to DataType then it has to be data right?

Reply:
The `params` also need to be convertible between pandas, numpy, spark, MLflow model server etc.

Reply:
This reminds me, @BenWilson2 @serena-ruan can we file a follow-up JIRA for the Productionization team to display the "params" signature section in the Model Registry UI? We don't have to own this, but it should definitely be on the Productionization team's radar :D

Reply:
Small tweak:

- For ``type``, only accept values defined in https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.DataType or strings referring to these values (see the type parameter of ColSpec: https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.ColSpec)

- Introduce a ``shape`` atribute, similar to TensorSpec. The acceptable values of ``shape`` should be 1 (scalar) and -1 (list / 1D array). We can expand this over time. If a ParamSpec has shape -1, we should only accept lists or 1-dimensional numpy arrays at inference-time. We can add support for multi-diimensional arrays later if we need them.

Reply:
+1 to the restrictions and type validations. If it helps, we should be validating any supported types are a subset of JSON supported types: https://en.wikipedia.org/wiki/JSON#Data_types. 
Can we add validation checks explicitly on the params input type to assure we're only whitelisting (and add tests that validate that parse checking passes with valid / fails for general Python objects, Python-specific data structures (i.e., tuples, DataClasses, Enums, etc.) so that we have a future-guarding protection against changes in the implementation over time?